core/src/main/java/com/linecorp/armeria/server/conditional/ConditionalRequestUtil.java

```diff
@@ -0,0 +1,209 @@
+        final List<ETag> requestETags = parseHeader(header);
+        if (requestETags != null) {
+            return ifMatch(requestETags, dataETag);
+        }
+        return PERFORM_METHOD;
+    }
+
+    /**
+     * Performs an If-None-Match.
+     * @param requestETags The parsed If-None-Match header.
+     * @param dataETag The ETag of the current data. If the resource doesn't exists, this is null.
+     *      This is mostly used to make sure PUT operations don't overwrite existing data.
+     * @return true if the method should be performed (ie. NOT just return 304 Not Modified).
+     */
+    public static ETagResponse ifNoneMatch(List<ETag> requestETags, @Nullable ETag dataETag) {
+        requireNonNull(requestETags, "requestETags cannot be null");
+        // Making sure that it's the same instance
+        if (requestETags == ETag.ASTERISK_ETAG) {
+            if (dataETag != null) {
+                return SKIP_METHOD_NOT_MODIFIED;
+            } else {
+                return PERFORM_METHOD;
+            }
+        }
+        if (dataETag != null) {
+            if (weakComparison(requestETags, dataETag)) {
+                return SKIP_METHOD_NOT_MODIFIED;
+            }
+        }
+        return PERFORM_METHOD;
+    }
+
+    /**
+     * Performs an If-None-Match.
+     * @param header The If-None-Match header.
+     * @param dataETag The ETag of the current data. If the resource doesn't exists, this is null.
+     *      This is mostly used to make sure PUT operations don't overwrite existing data.
+     * @return true if the method should be performed (ie. NOT just return 304 Not Modified).
+     */
+    public static ETagResponse ifNoneMatch(@Nullable String header, @Nullable ETag dataETag) {
+        final List<ETag> requestETags = parseHeader(header);
+        if (requestETags != null) {
+            return ifNoneMatch(requestETags, dataETag);
+        }
+        return PERFORM_METHOD;
+    }
+
+    /**
+     * Processes a request doing If-Match, If-None-Match, If-Unmodified-Since and If-Modified-Since.
+     * @param reqHeaders The RequestHeaders object of the request.
+     * @param dataETag The ETag of the current data. If the resource doesn't exists, this is null.
+     *      This is mostly used to make sure PUT operations don't overwrite existing data.
+     * @return HttpStatus to return. PRECONDITION_FAILED for when If-Match matches.
+     */
+    public static ETagResponse conditionalRequest(RequestHeaders reqHeaders,
+                                                  @Nullable ETag dataETag,
+                                                  @Nullable Long dataLastModified) {
+        requireNonNull(reqHeaders, "reqHeaders cannot be null");
+        final List<ETag> ifMatchHeader = parseHeader(reqHeaders.get(HttpHeaderNames.IF_MATCH));
+        if (ifMatchHeader != null && ifMatch(ifMatchHeader, dataETag) == SKIP_METHOD_PRECONDITION_FAILED) {
+            return SKIP_METHOD_PRECONDITION_FAILED;
```

Comment:
Shouldn't we return 412 according to the RFC?
```
1.  When recipient is the origin server and If-Match is present,
       evaluate the If-Match precondition:

       *  if true, continue to step 3

       *  if false, respond 412 (Precondition Failed) unless it can be
          determined that the state-changing request has already
          succeeded (see [Section 3.1](https://datatracker.ietf.org/doc/html/rfc7232#section-3.1))
```

Reply:
Shoudln't we check `If-Unmodified-Since`  after `if-match` check?
```
2.  When recipient is the origin server, If-Match is not present, and
       If-Unmodified-Since is present, evaluate the If-Unmodified-Since
       precondition:

       *  if true, continue to step 3

       *  if false, respond 412 (Precondition Failed) unless it can be
          determined that the state-changing request has already
          succeeded (see [Section 3.4](https://datatracker.ietf.org/doc/html/rfc7232#section-3.4))
```